### PR TITLE
ingest: translate known OpenTelemetry JVM metrics

### PR DESCRIPTION
**Pull request by axw**
## Motivation/summary

Copy well-known JVM metrics from their OpenTelemetry
names and associated labels to their counterparts
produced by the Elastic APM Java agent. This enables
users to visualise the metrics in the Elastic APM
app in Kibana.

Metrics are duplicated to avoid surprising users by
silently dropping the OpenTelemetry names. Users can
drop either one of the duplicates with ingest pipeline
customisation.

## Checklist

- [x] Update [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)
- [x] Documentation has been updated

## How to test these changes

1. Run apm-server
2. Instrument a Java application with https://github.com/open-telemetry/opentelemetry-java-instrumentation, pointed at apm-server
3. Wait for metrics to be collected and sent to APM Server (not sure how long this is)
4. Navigate to the APM app in Kibana, and check the GC and memory usage graphs are populated

See also https://github.com/elastic/apm-server/issues/4919

## Related issues

Closes https://github.com/elastic/apm-server/issues/4919

### CODE REVIEW

**Comment by axw** [comment]
@bmorelli25 I haven't added any docs here, not sure where would be appropriate?

**Review by felixbarny — COMMENTED** [review]
Why does the ingest pipeline get applied to all data streams? I'd expect it to only be in `/data_stream/app_metrics/elasticsearch/ingest_pipeline/apm_opentelemetry_metrics.json`

**Comment by bmorelli25** [comment]
> @bmorelli25 I haven't added any docs here, not sure where would be appropriate?

As this is a new default ingest pipeline, I'd like to see docs added here: https://github.com/elastic/apm-server/blob/master/docs/configuring-ingest.asciidoc. Could you add a new entry to the table that includes an added badge: `added::[7.13]`.

**Comment by axw** [comment]
> Could you add a new entry to the table that includes an added badge: added::[7.13].

Thanks @bmorelli25, will do.

> Why does the ingest pipeline get applied to all data streams? I'd expect it to only be in /data_stream/app_metrics/elasticsearch/ingest_pipeline/apm_opentelemetry_metrics.json

@felixbarny that would be ideal. With the old way of doing things, we only ever have a single pipeline. While we maintain both the old and new ways in tandem, we're just copying the single pipeline to all data streams in the integration package. In the future it would make sense to be more targeted.

**Comment by axw** [comment]
@bmorelli25 tagging you for review on the docs change. I also added an entry for another recently introduced ingest pipeline. Not sure how much of a pain it would be, but perhaps we should document the specific mapping when we update the OTel docs. Then in the ingest pipeline docs page we can link there.

File: docs/configuring-ingest.asciidoc [hunk 1]
```diff
@@ -40,6 +40,17 @@ and then viewing `ingest/pipeline/definition.json`.
 
 To disable this, or any other pipeline, set `output.elasticsearch.pipeline: _none`.
 
+`apm_error_grouping_name`::
```
Comment (bmorelli25) [inline docs/configuring-ingest.asciidoc:43]: L38-41 snuck in the middle of the table. Can you move L38-41 below the new content that you've added.

**Review by bmorelli25 — COMMENTED** [review]
> Not sure how much of a pain it would be, but perhaps we should document the specific mapping when we update the OTel docs. Then in the ingest pipeline docs page we can link there.

👍 Updated the doc issue accordingly.

File: docs/configuring-ingest.asciidoc (hunk 1 above)
Comment (axw) [reply docs/configuring-ingest.asciidoc:38]: Oops - fixed, thanks.

**Review by bmorelli25 — COMMENTED** [review]
Docs look good! Thanks.

**Comment by felixbarny** [comment]
> we're just copying the single pipeline to all data streams in the integration package. 

Is that a limitation in APM Server or in data streams?

> In the future it would make sense to be more targeted.

Is there an issue to track that?

**Comment by axw** [comment]
>> we're just copying the single pipeline to all data streams in the integration package.
>
> Is that a limitation in APM Server or in data streams?

APM Server. We're doing this for our own sanity, having a single source of truth for the pipeline.

>> In the future it would make sense to be more targeted.
>
> Is there an issue to track that?

There wasn't, but I've just now added an item to https://github.com/elastic/apm-server/issues/2631.

**Review by felixbarny — APPROVED** [review]

**Comment by axw** [comment]
jenkins run the tests please

File: ingest/pipeline/definition.yml [hunk 2]
```diff
@@ -74,3 +77,35 @@ apm_error_grouping_name:
...
+      # Copy `runtime.jvm.gc.count` (OpenTelemetry) to `jvm.gc.count` (Elastic APM).
+      field: jvm.gc.count
+      copy_from: runtime.jvm.gc.count
+      if: ctx.runtime?.jvm?.gc?.count != null
+  - set:
+      # Copy `runtime.jvm.gc.collection` (OpenTelemetry) to `jvm.gc.time` (Elastic APM).
+      # Both are defined in milliseconds. This is the old name for `runtime.jvm.gc.time`.
+      field: jvm.gc.time
+      copy_from: runtime.jvm.gc.collection
+      if: ctx.runtime?.jvm?.gc?.collection != null
```
Comment (simitt) [inline ingest/pipeline/definition.yml:105]: nit: It probably shouldn't happen, but could `runtime.jvm.gc.time` and `runtime.jvm.gc.collection` both be set? If so the value from the _old_ name would override the newer one. I'd probably flip it, or set `override: false` here.

File: ingest/pipeline/definition.yml [hunk 3]
```diff
@@ -74,3 +77,35 @@ apm_error_grouping_name:
...
+  - set:
+      # Copy `labels.gc` (OpenTelemetry) to `labels.name` (Elastic APM), for jvm.gc.{time,count}.
+      field: labels.name
+      copy_from: labels.gc
+      override: false # don't replace existing labels.name field, if any
+      if: ctx.labels?.gc != null
```
Comment (simitt) [inline ingest/pipeline/definition.yml:111]: This is not only applied for `jvm.gc.{time,count}` but to every `labels.gc` right?

File: apmpackage/apm/0.1.0/data_stream/app_metrics/elasticsearch/ingest_pipeline/apm_opentelemetry_metrics.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
+{
+  "description": "Populate Elastic APM metric fields from well-known OpenTelemetry metric counterparts",
+  "processors": [
+    {
+      "set": {
+        "field": "jvm.memory.{{labels.area}}.{{labels.type}}",
+        "copy_from": "runtime.jvm.memory.area",
+        "if": "ctx.runtime?.jvm?.memory?.area != null && ctx.labels?.area != null && ctx.labels?.type != null"
```
[the remark this answers is not in the document]
Comment (axw) [reply apmpackage/apm/0.1.0/data_stream/app_metrics/elasticsearch/ingest_pipeline/apm_opentelemetry_metrics.json:8]: No, hence the use of the [null safe](https://www.elastic.co/guide/en/elasticsearch/painless/current/painless-operators-reference.html#null-safe-operator) operator. If `labels` is null, then `ctx.labels?.area` will evaluate to `null`.

File: ingest/pipeline/definition.yml (hunk 2 above)
Comment (axw) [reply ingest/pipeline/definition.yml:105]: Fair, I'll move `runtime.jvm.gc.collection` above `runtime.jvm.gc.time`.

File: ingest/pipeline/definition.yml (hunk 3 above)
Comment (axw) [reply ingest/pipeline/definition.yml:111]: Indeed, probably a bit too lax. I'll add `&& ctx.runtime?.jvm?.gc != null` to the condition.

**Comment by axw** [comment]
run elasticsearch-ci/docs

**Comment by axw** [comment]
jenkins run the tests please

**Comment by simitt** [comment]
Tested with BC4:
Instrumented a java test app with the otel instrumentation and waited until metrics documents existed. No metrics show up in the APM UI. 

Screenshot 2021-05-12 at 15 02 26

Screenshot 2021-05-12 at 15 00 11

**Comment by axw** [comment]
> agent.name needs to be java (not opentelemetry/java) so that the JVM tab shows up

Ugh :(
I don't think we should be changing the agent name, so this part will probably need to happen in the UI.

> minimum jvm.memory.heap.used needs to exist for JVM details to be shown.

Hmm. Is there just one metric document? We should be capturing `jvm.memory.heap.used`. Maybe what's happening is that we're splitting the metrics into multiple docs - would that explain this?

**Comment by felixbarny** [comment]
What might also be missing is `agent.ephemeral_id`. IIRC, the UI needs that for downsampling/calculating derivatives for the GC time and count.

**Comment by simitt** [comment]
created https://github.com/elastic/kibana/issues/100253 as a follow up for showing opentelemetry JVM metrics. 

> Hmm. Is there just one metric document? We should be capturing jvm.memory.heap.used. Maybe what's happening is that we're splitting the metrics into multiple docs - would that explain this?

**Comment by simitt** [comment]
created a follow up for the metrics docs https://github.com/elastic/apm-server/issues/5275

**Comment by axw** [comment]
> What might also be missing is agent.ephemeral_id. IIRC, the UI needs that for downsampling/calculating derivatives for the GC time and count.

@felixbarny there's a TypeScript definition for `ephemeral_id`, but no other code in the UI that actually references the field in aggs or otherwise. Does that indicate a UI bug?

**Comment by felixbarny** [comment]
> Does that indicate a UI bug?

Looking at the aggs that Kibana does, I think it doesn't work as I'd expect

https://github.com/elastic/kibana/blob/1f02c48d3cf57dd921cfe64186a695f8375e3954/x-pack/plugins/apm/server/lib/metrics/by_agent/java/gc/fetch_and_transform_gc_metrics.ts#L70-L109

It does a derivative over the max value of gc.count across instances as opposed to summing the derivatives of each instance.

I guess it would be easier for the UI if the Java agent reported deltas for gc.time instead of an incrementing counter. However, it might be very tricky to adjust the OTel metrics to report deltas.
IIRC, there's some setting in OTel that allows to report deltas instead of monotonically incrementing counters. But I'm not sure if that setting also applies for the [GC metrics that are captured via a `longSumObserver`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/runtime-metrics/library/src/main/java/io/opentelemetry/instrumentation/runtimemetrics/GarbageCollector.java).

**Comment by simitt** [comment]
@felixbarny do you mind moving this over to a new issue? It sounds like it'd be worth a follow up discussion.

**Comment by felixbarny** [comment]
After giving it another thought, it's probably not an issue as that chart is only ever shown in the context of a specific instance. Therefore, the ephemeral_id is not needed.

**Pull request merged**